src/v/archival/archiver_manager.cc

```diff
@@ -0,0 +1,1051 @@
+    //  group.
+    struct st_stopping_async : state<managed_partition_state_t::stopping> {
+        std::optional<ssx::semaphore_units> reset_units;
+
+        void clear() noexcept { reset_units = std::nullopt; }
+    };
+
+    // FSM will be created with this state
+    using initial_state = st_passive;
+
+    // List of events
+    struct ev_leadership_acquired
+      : managed_partition_event_base<
+          managed_partition_event_t::leadership_acquired,
+          ev_leadership_acquired> {
+        model::term_id term;
+    };
+    struct ev_archiver_started
+      : managed_partition_event_base<
+          managed_partition_event_t::archiver_started,
+          ev_archiver_started> {
+        // started archiver
+        ss::lw_shared_ptr<ntp_archiver> archiver;
+    };
+    struct ev_archiver_failure
+      : managed_partition_event_base<
+          managed_partition_event_t::archiver_failure,
+          ev_archiver_failure> {
+        std::exception_ptr error;
+    };
+    struct ev_leadership_lost
+      : managed_partition_event_base<
+          managed_partition_event_t::leadership_lost,
+          ev_leadership_lost> {};
+    struct ev_archiver_stopped
+      : managed_partition_event_base<
+          managed_partition_event_t::archiver_stopped,
+          ev_archiver_stopped> {};
+    struct ev_shutdown
+      : managed_partition_event_base<
+          managed_partition_event_t::shutdown,
+          ev_shutdown> {};
+
+    // List of transition functors.
+    // The functors are named tr_(source-state)_to_(target-state).
+    // The source-state is a state from which we're transitioning and the
+    // target-state is the final state. Each functor implements operator () for
+    // the combination of source state, target state and the triggering event.
+    // The functor is added to the transition table and parameters of the
+    // call-operator should match the types in the row of the table.
+    struct tr_passive_to_starting {
+        void operator()(
+          const ev_leadership_acquired& new_leadership,
+          state_machine_t& fsm,
+          st_passive& prev,
+          st_starting_async& next) {
+            vlog(fsm._ctxlog.info, "starting ntp_archiver {}", new_leadership);
+            prev.clear();
+            next.reset_units = fsm._part->get_archiver_reset_units();
+            vassert(
+              next.reset_units.has_value(), "Reset units are not acquired");
+            auto completion =
+              [&fsm](ss::future<ss::lw_shared_ptr<ntp_archiver>> fut) {
+                  if (fut.failed()) {
+                      fsm.process_event(
+                        ev_archiver_failure{.error = fut.get_exception()});
+                  } else {
+                      fsm.process_event(
+                        ev_archiver_started{.archiver = fut.get()});
+                  }
+              };
+            ssx::background = fsm.create_and_start_archiver(new_leadership.term)
```

Comment:
We background the future without any `ss::gate`s held?

Reply:
Yes, the lifetime is guaranteed by the FSM. We only disposing the object when it's in the `st_passive` state.

Reply:
Follow up question: who is handling the exception returned by `create_and_start_archiver`?